Sensors/SMCProcessor/SMCProcessor.cpp

```diff
@@ -56,9 +61,112 @@ void SMCProcessor::readRapl() {
 	}
 }
 
-void SMCProcessor::updateCounters() {
-	uint32_t cpu = cpu_number();
+IOReturn SMCProcessor::bindCurrentThreadToCpu(uint32_t cpu)
+{
+	// Obtain power management callbacks
+	pmCallBacks_t callbacks {};
+	pmKextRegister(PM_DISPATCH_VERSION, nullptr, &callbacks);
+	
+	if (!callbacks.LCPUtoProcessor) {
+		SYSLOG("scpu", "failed to obtain LCPUtoProcessor");
+		return KERN_FAILURE;
+	}
+	
+	if (!callbacks.ThreadBind) {
+		SYSLOG("scpu", "failed to obtain ThreadBind");
+		return KERN_FAILURE;
+	}
+	
+	auto preemptionLock = IOSimpleLockAlloc();
+	if (!preemptionLock) {
+		SYSLOG("scpu", "Preemption lock cannot be allocated");
+		return KERN_FAILURE;
+	}
+	
+	if (!IOSimpleLockTryLock(preemptionLock)) {
+		SYSLOG("scpu", "Preemption cannot be disabled before performing ThreadBind");
+		IOSimpleLockFree(preemptionLock);
+		return KERN_FAILURE;
+	}
+	
+	bool success = true;
+	auto enable = ml_set_interrupts_enabled(FALSE);
+	
+	while (1)
+	{
+		auto processor = callbacks.LCPUtoProcessor(cpu);
+		if (processor == nullptr) {
+			SYSLOG("scpu", "failed to call LCPUtoProcessor with cpu %u", cpu);
+			success = false;
+			break;
+		}
+		callbacks.ThreadBind(processor);
+		break;
+	}
+
+	IOSimpleLockUnlock(preemptionLock);
+	ml_set_interrupts_enabled(enable);
+	
+	IOSimpleLockFree(preemptionLock);
+
+	return success ? KERN_SUCCESS : KERN_FAILURE;
+}
 
+void SMCProcessor::staticThreadEntry(thread_call_param_t param0, thread_call_param_t param1)
+{
+	auto *that = OSDynamicCast(SMCProcessor, reinterpret_cast<OSObject*>(param0));
+	if (!that) {
+		SYSLOG("scpu", "Failed to get pointer to SMIMonitor");
+		return;
+	}
+	
+	uint32_t cpu = static_cast<uint32_t>(reinterpret_cast<uint64_t>(param1));
+	// This should not happen
+	if (cpu >= CPUInfo::MaxCpus) {
+		SYSLOG("scpu", "CPU number cannot exceed %u", CPUInfo::MaxCpus);
+		return;
+	}
+	
+	DBGLOG("scpu", "staticThreadEntry for CPU number %u is started", cpu);
+
+	bool success = true;
+	while (1) {
+			
+		IOReturn result = that->bindCurrentThreadToCpu(cpu);
+		if (result != KERN_SUCCESS) {
+			success = false;
+			break;
+		}
+		
+		IOSleep(20);
+		
+		auto enable = ml_set_interrupts_enabled(FALSE);
```

Comment:
Is it really worth disabling interrupts when calling this or in fact doing this check at all? The kernel may switch to any CPU at any time, and disabling interrupts will not change that. To be honest, just trust that we did switch. Maybe add an assert(cpu_number() == cpu); in the debug code at most.